IPv6 support

KeeJef opened the pull request:
Adds IPv6 support for daemon p2p, daemon RPC, and wallet RPC.

Currently there is no option to bind different ports for
IPv4 and IPv6 for RPC, but that can be changed if necessary.

References this issue https://github.com/monero-project/monero/issues/147

TheCharlatan commented on contrib/epee/include/net/net_utils_base.h:
```diff
@@ -89,6 +89,48 @@ namespace net_utils
+		constexpr bool is_same_host(const ipv6_network_address& other) const noexcept
+		{ return ip() == other.ip(); }
+
+		constexpr std::string ip() const noexcept { return m_ip; }
```
You should not use std::string in a constexpr , constexpr only takes literal types. Travis and build bots catch this when compiling with -Winvalid-constexpr.

TheCharlatan commented on contrib/epee/include/net/net_utils_base.h:
```diff
@@ -89,6 +89,48 @@ namespace net_utils
+		bool equal(const ipv6_network_address& other) const noexcept;
+		bool less(const ipv6_network_address& other) const noexcept;
+		constexpr bool is_same_host(const ipv6_network_address& other) const noexcept
+		{ return ip() == other.ip(); }
```
This trips -Winvalid-constexpr. To me it looks like it's because they have different lifetimes, which is not permitted.

moneromooo-monero commented:
I've not really looked yet, but why are you passing two sets of host/port ?
Also, I'm not sure we want IPv6 support, as the scarceness of IPv4 helps defend against Sybils somewhat.

SamsungGalaxyPlayer commented:
> Also, I'm not sure we want IPv6 support, as the scarceness of IPv4 helps defend against Sybils somewhat.

Not sure if I agree with this being a consideration to reasonably worry about.

KeeJef commented:
> I've not really looked yet, but why are you passing two sets of host/port ?

for whatever reason, someone wants to bind to different ports for IPv4 and IPv6

KeeJef commented:
Regarding the usefulness of the change i think this has been a pretty highly requested feature for a while, i know personally of a couple of people running IPv6 only networks who struggle to use Monero as is.

KeeJef commented:
Will respond to @moneromooo-monero code review shortly, have added a commit to turn off IPv6 by default, this should address concerns over Sybil attacks while still allowing IPv6 full nodes to function. 

Also changed resolution order to IPv4 first then falling back on IPv6, and fixed some bugs we found when we deployed this.

moneromooo-monero commented:
FYI, the proxy patch (https://github.com/monero-project/monero/pull/4988) is pretty much ready and will hopefully be merged soon so you'll have to rebase onto that. Might be significant changes.

moneromooo-monero commented:
Do you want to rebase ? I don't think we're expecting major changes in the net code (except maybe i2p stuff, but I'm not sure what anonimal is doing atm so you could get in first).

KeeJef commented:
Yeah sorry, we will get to this shortly and rebase, hopefully within the next 2 weeks

moneromooo-monero commented:
Actually, https://github.com/monero-project/monero/pull/5090 is still waiting for a rebase before merging, but that's the last one I'm expecting.

KeeJef commented:
This should be rebased now @moneromooo-monero also fixed up a few of the issues you commented on

moneromooo-monero commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -983,15 +1058,23 @@ PRAGMA_WARNING_DISABLE_VS(4355)
+      MERROR("Failed to convert port no = " << port_ipv6);
+      return false;
+    }
+    return this->init_server(p, address, p_ipv6, address_ipv6, use_ipv6, require_ipv4, ssl_options);
```
I'm guessing the std::move being removed was a merge error.

tewinget replied:
correct, fixed.

moneromooo-monero commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1149,29 +1232,52 @@ POP_WARNINGS
+      (*current_new_connection).reset(new connection<t_protocol_handler>(io_service_, m_state, m_connection_type, conn->get_ssl_support()));
+      current_acceptor->async_accept((*current_new_connection)->socket(),
+          boost::bind(accept_function_pointer, this,
+            boost::asio::placeholders::error));
```
Can you fix the reformatting please ? Makes it hard to see what changed.

tewinget replied:
The reformatting is because the rest of the file is spaces and this bit was tabs, or so it would seem.  It can be changed back to tabs if necessary, but it seems like all of the non-whitespace changes are highlighted properly in the diff.

moneromooo-monero replied:
Fair enough, I guess half the lines are changed anyway.

tewinget replied:
That was my thought too, though the changes are small-ish.  Previously for this PR you had asked for less code duplication and this was the cleanest way (without significant refactoring) I could see to have that.

moneromooo-monero commented on contrib/epee/include/net/local_ip.h:
```diff
@@ -27,10 +27,27 @@
+    bool is_ipv6_local(const std::string& ip)
+    {
+      // ipv6 local addresses start with fc00/7 -- (fcXX or fdXX)
+      return (ip.find("fc") == 0 || ip.find("FC") == 0 || ip.find("fd") == 0 || ip.find("FD") == 0);
```
Probably want to match fC etc too. boost::algorithm::iequals does it.

tewinget replied:
addressed.

moneromooo-monero commented on contrib/epee/include/net/local_ip.h:
```diff
@@ -27,10 +27,27 @@
+    bool is_ipv6_loopback(const std::string& ip)
+    {
+      // ipv6 loopback is ::1
+      return (ip == "::1");
```
That's just one short form though isn't it ?

tewinget replied:
switched to using boost's address_v6::is_loopback()

moneromooo-monero commented on contrib/epee/include/net/net_utils_base.h:
```diff
@@ -107,6 +107,54 @@ namespace net_utils
 
+	class ipv6_network_address
+	{
+		std::string m_ip;
```
Why std::string, rather than an array of however number of bytes ? The string means it's never certain to be canonical.

tewinget replied:
changed to using boost's address_v6 under the hood

moneromooo-monero replied:
Looks good, any chance the m_address_str that's still left can be made canonical ? Otherwise stuff that uses the host string as a key (ie, the block list) will still see the same address as different.

tewinget replied:
It looks a bit weird, but yes (and it should be fine despite the appearance), and done.

moneromooo-monero commented on contrib/epee/src/net_utils_base.cpp:
```diff
@@ -21,6 +21,19 @@ namespace epee { namespace net_utils
+	{ return is_same_host(other) && port() == other.port(); }
+
+	bool ipv6_network_address::less(const ipv6_network_address& other) const noexcept
+	{ return is_same_host(other) ? port() < other.port() : ip() < other.ip(); }
```
Those will consider equal IPs different AFAICT, given the non canonical nature of the std::string ip field.

tewinget replied:
changed to using boost's address_v6 under the hood

moneromooo-monero commented on src/net/parse.cpp:
```diff
@@ -34,28 +34,67 @@
+                port = address.substr(host.size() + 1);
+            }
+        }
+    }
```
That seems way underspecified.

tewinget replied:
Agreed that it seems that way, but it's just a helper function to avoid a bit of code duplication; the places it is called will handle making sure what it spits out is correct/usable.

moneromooo-monero commented on src/net/parse.cpp:
```diff
@@ -34,28 +34,67 @@
+        get_network_address_host_and_port(std::string(address), host_str, port_str);
+
+        boost::string_ref host_str_ref(host_str);
+        boost::string_ref port_str_ref(host_str);
```
That one seems wrong.

tewinget replied:
correct, fixed.

moneromooo-monero commented on src/net/parse.cpp:
```diff
@@ -34,28 +34,67 @@
+        if (host_str_ref.ends_with(".onion"))
             return tor_address::make(address, default_port);
-        if (host.ends_with(".i2p"))
+        if (host_str_ref.ends_with(".i2p"))
```
This all seems to be gratuituous changes, no ?

tewinget replied:
Seems so, but was done in order to pull get_network_address_host_and_port() (see above) into its own function, as the same logic was used elsewhere in the codebase and should not be duplicated if possible.

moneromooo-monero commented on src/rpc/rpc_args.cpp:
```diff
@@ -81,6 +90,34 @@ namespace cryptonote
+      // allow square braces, but remove them here if present
+      if (config.bind_ipv6_address.find('[') != std::string::npos)
+      {
+        config.bind_ipv6_address = config.bind_ipv6_address.substr(1, config.bind_ipv6_address.size() - 2);
```
That seems to remove any last char, regardless of whether ] or not.

tewinget replied:
If the address was specified with an opening square brace at the beginning and not a closing square brace at the end then it was specified incorrectly.  I reasoned that it would be incredibly tedious to look for all of the possible permutations of incorrect address specification and give hints (e.g. "missing ] character after address") as there are quite a few ways of messing that up.  In this case if the ] is missing then when the code tries to use the address it will fail and message/error accordingly.

moneromooo-monero commented on contrib/epee/include/net/net_utils_base.h:
```diff
@@ -107,6 +108,73 @@ namespace net_utils
+
+			post_deserialize<is_store>(this_ref);
+		END_KV_SERIALIZE_MAP()
+	};
```
That all seem dangerous. Is it possible to just remove that m_address_str, and just have the ipv6 var ? host_str calls to_string. No need for this post_deserialize.

tewinget replied:
Would it be preferable to use boost's address_v6 to make the string address canonical in the case where the address is passed to the constructor and trust that if we're deserializing one it is canonical and that address_v6.to_string() does indeed return at least a consistent (if not canonical, consistent should be fine) string, or to store the address bytes?  I can use the underlying representation address_v6 uses and that should work nicely, but I'll have to add std::array to the list of stl containers supported by KV_SERIALIZE.

moneromooo-monero replied:
I'm not sure I understand that :)

- having a string seems wrong since several strings can represent the same address
- I assume the boost address_v6 is just 8 bytes and generally behaves sensibly; if not you can just make your own
- For serialization, you can save as a blob assuming the boost object is not a complex thing, which it has no reason to be, but I'm guessing here

tewinget replied:
re: strings, I agree that that's not the best way to do things, though as I said if address_v6::to_string() is consistent in how it represents an address then there shouldn't be an issue of multiple strings for the same address.

address_v6 uses a std::array of 16 bytes internally.  It does behave sensibly but again I'd need to add it to the list of containers supported by KV_SERIALIZE.  This seems like the best approach, I just thought I'd confirm as much.

The problem with serializing as a different representation (e.g. blob it up before serializing) is that on deserialize the underlying field must be populated, hence the post_serialize above.  I agree that this approach feels clumsy; at the time I hadn't realized that address_v6 exposes the raw bytes, simply misread the header.

I'll go ahead and use the std::array as the underlying representation as it seems like the cleanest approach at this point.

moneromooo-monero replied:
Can you not use KV_SERIALIZE_POD_AS_BLOB ?

tewinget replied:
Ok, nevermind, KV_SERIALIZE did not want to handle std::array and if it's not going to be that simple to tell it to it's probably not worth it.

So the issue at hand is that doing a "post-serialize" function is a bit messy.  I agree with that, but the alternative seems to be running code every time one of the access functions is used, either to check if the struct has been updated since being deserialized or to do what that updating would do every time.

That seems worse, so I'd say it's between post-serialize or trust that address_v6::to_string() will always give the same string for a given address.  It's possible I'm missing something obvious as I am rather quite tired; I'm genuinely not trying to be obtuse.  Given that serializing std::array wasn't a quick win and post-serialize is arguably ugly, I'm inclined to say trust that address_v6::to_string() is consistent.

tewinget replied:
Should have hit refresh.  Will take a look at KV_SERIALIZE_POD_AS_BLOB, not sure if that will sort it.

moneromooo-monero replied:
https://paste.debian.net/hidden/f4a22592/ are the changes I'd make. This uses only a address_v6 object. I think there might even be fewer conversions overall. I've only built it, not tried it.

tewinget replied:
Thanks for that.  There were a couple issues to touch up with that patch but no worries.  Biggest is that address_v6.from_string is a static member function.  Why the compiler allows instance.static_function() is beyond me, but finding that sure was fun.

moneromooo-monero replied:
That const_cast should not be needed, but actually was when I tried it. Not sure why... I might look at that later too.

tewinget commented:
PR updated again.  Adapted @moneromooo-monero 's patch for the ipv6 network_address representation as well as fixing some peerlist related issues I found in the meantime.

moneromooo-monero commented on contrib/epee/include/net/abstract_tcp_server2.inl:
```diff
@@ -1203,10 +1309,10 @@ POP_WARNINGS
-    acceptor_.async_accept(new_connection_->socket(),
-      boost::bind(&boosted_tcp_server<t_protocol_handler>::handle_accept, this,
-      boost::asio::placeholders::error));
+    (*current_new_connection).reset(new connection<t_protocol_handler>(io_service_, m_state, m_connection_type, (*current_new_connection)->get_ssl_support()));
```
Do you actually need a new_connection_ipv6 ? It looks like the same object can be used for both.

tewinget replied:
If we require IPv4 and IPv6 use the same port then yes, I believe it could.  I don't think it's a good idea to have two async_accept (one for each protocol, in case they're bound differently) to be bound to the same new_connection object, but I may be mistaken.

moneromooo-monero replied:
OK, two is fine then.

moneromooo-monero commented on contrib/epee/include/net/local_ip.h:
```diff
@@ -27,10 +27,33 @@
+      static const std::string match_str1 = "fc";
+      static const std::string match_str2 = "fd";
+
+      return (boost::iequals(first_two, match_str1) || boost::iequals(first_two, match_str2));
```
I'd rather this used the boost from_string + is_link_local, since it turns out those exist.

tewinget replied:
I looked at that as well but I don't think is_link_local does exactly what we want.

https://github.com/boostorg/asio/blob/develop/include/boost/asio/ip/impl/address_v6.ipp#L168

for both is_link_local and is_site_local it's looking if the address is link-local, but I don't think that's what we care about.  I could be mistaken here, but I presume we care about unique-local-addresses in IPv6, i.e. the equivalent of 10.0.0.0/8 or 192.168/16

If link-local is indeed what we care about it can certainly be changed.

moneromooo-monero replied:
Looking at the docs, is_site_local is what we want here.

moneromooo-monero replied:
Actually, not sure. I see a claim that site-local allows talking to other machines on the LAN, which is what we want (if there is a "only"). So I don't get the difference between link and site then.

tewinget replied:
Well, so I'm not sure in practice which would matter.  Could even check for both.

I just know that the wikipedia article compared link-local in IPv6 as the equivalent of 169.254.0.0/16 in IPv4.  It also mentioned something the spec requiring every interface to have a link-local address, though I did not quite understand to what end.

I guess maybe just go with both?

moneromooo-monero replied:
I think we want site local. The use of it is to detect when a node is not a loopback node, but is still in the LAN, to default to trusted. Site local seems to encompass all the non routable IPv4 address ranges that are meant for LAN only use.

tewinget replied:
So looking into it on IANA's site, site-local is deprecated as of a 2004 RFC.

RFC 4193 states that Unique Local Unicast addresses are intended for use within a "site" and not intended to be routed on the internet.

RFC 4291 indicates that routers are not intended to forward packets from link-local addresses (fe80::/10) but I guess it could sorta still be used similarly to loopback, though that would be very strange indeed.

I shall add is_site_local to the function, but the current check for unique local unicast should remain.

moneromooo-monero replied:
OK. Then we'll just wait for someone with an IPv6 LAN to tell us whether it works as expected or not.

moneromooo-monero commented on contrib/epee/include/net/net_parse_helpers.h:
```diff
@@ -128,11 +128,51 @@ namespace net_utils
   }
 
+  inline
+    bool parse_url_ipv6(const std::string url_str, http::url_content& content)
```
That function makes me a bit uncomfortable, but I think it's fine to leave as is. I might come back later and see if it can be removed.

moneromooo-monero commented on contrib/epee/src/net_helper.cpp:
```diff
@@ -11,10 +11,39 @@ namespace net_utils
 		boost::asio::ip::tcp::resolver::iterator end;
-		if(iterator == end) // Documentation states that successful call is guaranteed to be non-empty
-			throw boost::system::system_error{boost::asio::error::fault, "Failed to resolve " + addr};
+    boost::system::error_code resolve_error;
```
Can I ask to indent this as the surrounding code ? :)

tewinget replied:
done

moneromooo-monero commented on src/net/parse.cpp:
```diff
@@ -34,28 +34,69 @@
+            if (host.size() < address.size())
+            {
+                port = address.substr(host.size() + 1);
+            }
```
This function can also use boost from_string

tewinget replied:
That function takes an address that may be specified as "address:port" and separates the two; address_v6.from_string and address_v4.from_string -- ok, they're deprecated, so make_address_vN() -- use inet_pton under the hood which would not play nice with the ":port" part nor the square braces around the address in the case of ipv6.

tewinget commented:
New comments addressed, commits squashed and pushed.

KeeJef commented:
Any updates here? looks like we have merge conflicts again.

selsta commented:
@KeeJef Everything got kinda delayed because of the v0.14.1.0 and v0.14.1.2 release and moneromooo is taking time off for the summer so merges will probably take a while :)

selsta commented:
@KeeJef Could you please rebase this PR?

dEBRUYNE-1 commented:
@KeeJef - If you can rebase this as soon as possible, we can merge this pull request next (in order to avoid you having to rebase a second time).

moneromooo-monero commented:
Yes, this can go in next as soon as it builds. Thanks for the patch.

KeeJef commented:
Should be ready to go again
